shared/db/queries/user.js

```diff
@@ -0,0 +1,693 @@
+// @flow
+import { createReadQuery, createWriteQuery, db } from 'shared/db';
+import { uploadImage } from 'api/utils/file-storage';
+import { createNewUsersSettings } from 'api/models/usersSettings';
+import { sendNewUserWelcomeEmailQueue } from 'shared/bull/queues';
+import { events } from 'shared/analytics';
+import { trackQueue, identifyQueue } from 'shared/bull/queues';
+import { removeUsersCommunityMemberships } from 'api/models/usersCommunities';
+import { removeUsersChannelMemberships } from 'api/models/usersChannels';
+import { disableAllThreadNotificationsForUser } from 'api/models/usersThreads';
+import { disableAllUsersEmailSettings } from 'api/models/usersSettings';
+import { getUserChannelIds } from 'api/models/usersChannels';
+import type { PaginationOptions } from 'api/utils/paginate-arrays';
+import type { DBUser, FileUpload, DBThread } from 'shared/types';
+
+export const getUserById = createReadQuery((userId: string) => ({
+  query: db.table('users').get(userId),
+  tags: (user: ?DBUser) => [userId],
+}));
+
+export const getUserByEmail = createReadQuery((email: string) => ({
+  query: db.table('users').getAll(email, { index: 'email' }),
+  process: (users: ?Array<DBUser>) => (users && users[0]) || null,
+  tags: (user: ?DBUser) => (user ? [user.id] : []),
+}));
+
+export const getUserByUsername = createReadQuery((username: string) => ({
+  query: db.table('users').getAll(username, { index: 'username' }),
+  process: (users: ?Array<DBUser>) => (users && users[0]) || null,
+  tags: (user: ?DBUser) => (user ? [user.id] : []),
+}));
+
+export const getUsersByUsername = createReadQuery(
+  (usernames: Array<string>) => ({
+    query: db.table('users').getAll(...usernames, { index: 'username' }),
+    tags: (users: ?Array<DBUser>) => (users ? users.map(({ id }) => id) : []),
```

Comment:
I believe this can error if `users` is an empty array. It will throw `cannot destructure property id of null`, or something like that - right? Should it be `users && users.length > 0 ? ...`?